src/main/kotlin/uk/gov/justice/digital/hmpps/sentenceplan/services/PlanService.kt

```diff
@@ -150,4 +152,63 @@ class PlanService(
     // make sure we update the previous version with the new status, not the new one.
     return planVersionRepository.save(plan)
   }
+
+  fun countersignPlan(planUuid: UUID, countersignPlanRequest: CounterSignPlanRequest): PlanVersionEntity {
+    val version = planVersionRepository.getVersionByUuidAndVersion(planUuid, countersignPlanRequest.sentencePlanVersion.toInt())
+
+    // Duplicate request checking
+    when (countersignPlanRequest.signType) {
+      CountersignType.COUNTERSIGNED -> {
+        if (version.status == CountersigningStatus.COUNTERSIGNED) {
+          throw ConflictException("Plan $planUuid was already countersigned.")
+        }
+      }
+      CountersignType.REJECTED -> {
+        if (version.status == CountersigningStatus.REJECTED) {
+          throw ConflictException("Plan $planUuid was already rejected.")
+        }
+      }
+      CountersignType.DOUBLE_COUNTERSIGNED -> {
+        if (version.status == CountersigningStatus.DOUBLE_COUNTERSIGNED) {
+          throw ConflictException("Plan $planUuid was already double countersigned.")
+        }
+      }
+      CountersignType.AWAITING_DOUBLE_COUNTERSIGN -> {
+        if (version.status == CountersigningStatus.AWAITING_DOUBLE_COUNTERSIGN) {
+          throw ConflictException("Plan $planUuid was already awaiting double countersign.")
+        }
+      }
+    }
+
+    // Valid transitions
+    when (countersignPlanRequest.signType) {
+      CountersignType.COUNTERSIGNED -> {
+        if (version.status != CountersigningStatus.AWAITING_COUNTERSIGN) {
+          throw ConflictException("Plan $planUuid was not awaiting countersign.")
+        }
+        version.status = CountersigningStatus.COUNTERSIGNED
```

Comment:
can these `version.status=` lines just be one line before the `.save` of `version.status = countersignPlanRequest.signType` ?

Reply:
oh, I see, they're not the same type. Should they be?

Reply:
I don't think so. `CountersigningStatus` has more that this endpoint needs to take.

Reply:
OK. I guess if we were desperate to optimise we could do something like `version.status = CountersigningStatus.valueOf(countersignPlanRequest.signType.name)`? 🤷